test/e2e/e2e_test.go

```diff
@@ -70,11 +79,72 @@ func TestMain(m *testing.M) {
 		cancel()
 	}()
 
+	if globalOpts.ArtifactDir != "" {
+		go setupMetricsEndpoint(log)
+		go e2eObserverControllers(testContext, log, globalOpts.ArtifactDir)
+		defer dumpTestMetrics(log, globalOpts.ArtifactDir)
+	}
+
 	// Everything's okay to run tests
 	log.Info("executing e2e tests", "options", globalOpts)
 	os.Exit(m.Run())
 }
 
+func e2eObserverControllers(ctx context.Context, log logr.Logger, artifactDir string) {
+	mgr, err := ctrl.NewManager(e2eutil.GetConfigOrDie(), manager.Options{MetricsBindAddress: "0"})
+	if err != nil {
+		log.Error(err, "failed to construct manager for observers")
+		return
+	}
+	if err := podtimingcontroller.SetupWithManager(mgr, log, artifactDir); err != nil {
+		log.Error(err, "failed to set up podtimingcontroller")
+		return
+	}
+
+	if err := mgr.Start(ctx); err != nil {
+		log.Error(err, "Mgr ended")
+	}
+}
+
+const metricsServerAddr = "127.0.0.1:8080"
+
+func setupMetricsEndpoint(log logr.Logger) {
+	log.Info("Setting up metrics endpoint")
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.Handler())
+	server := &http.Server{Addr: metricsServerAddr, Handler: mux}
+	if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+		log.Error(err, "metrics server ended unexpectedly")
+	}
```

Comment:
Can we set this up to use the test context for graceful shutdown? Like....

```go
mux := http.NewServeMux()
mux.Handle("/metrics", promhttp.Handler())
server := http.Server{
  Addr:         opts.MetricsAddr,
  Handler:      mux,
  ReadTimeout:  5 * time.Second,
  WriteTimeout: 10 * time.Second,
}
go func() {
  <-ctx.Done()
  if err := server.Shutdown(ctx); err != nil {
    log.Error(err, "failed to shut down metrics server")
  }
}()
if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
  log.Error(err, "metrics server closed unexpectedly")
}

```

Reply:
Actually we can't do that, because the `defer` we use for scraping might end up getting a connection refused